Adding DecaNLP  (updated #1231)

hadasah opened the pull request:
@CharlieChenFB added the DecaNLP tasks. I've updated his PR so that the CNN_DM, MWSC, and SST datasets w/ train/dev/test splits match up with the DecaNLP paper and repo. 
Two notable discrepancies with the DecaNLP tasks remain:  
1) IWSLT in the DecaNLP task uses 2016 for train, 2013 and 2014 test sets for dev and test, resp. I just kept IWSLT 2014 for now, we can add a DecaNLP specific IWSLT task if we feel strongly. 
2) WOZ # examples in the paper isn't what we get when running display_data. Based on the repo, it seems they might use both English and German data despite reporting only using English in the paper. I'm keeping just English for and initial release, display_data output looks reasonable, we can revisit if this is important.

Only adding reviewers who responded to the original PR for now.

CharlieChenFB commented:
Hi! Thank you for picking this up, and sorry that I wasn't able to close this off during Bootcamp.

A few things -- one problem I encountered was that the tests might time out (and take a very long time) since this is such a large dataset. Also, I tried to fix a recursion problem in the multitask teacher, which occurs if you try to run, for example, babi:MultiTaskTeacher.

Finally, there was a small problem with the linter disagreeing with the escaped square brackets.

stephenroller commented on parlai/core/agents.py:
```diff
@@ -682,7 +682,7 @@ def create_task_agent_from_taskname(opt):
     if not opt.get('task'):
         opt['task'] = 'pytorch_teacher'
-    if ',' not in opt['task']:
+    if ',' not in opt['task'] and 'MultiTaskTeacher' not in opt['task']:
```
Can you describe the setting that this workaround fixes? It feels like a bandaid, rather than a root cause.

CharlieChenFB replied:
Yeah, I agree. That this is a bandaid instead of a fix -- What should be the expected behaviour of calling <task_name>:MultiTaskTeacher?

hadasah replied:
Reverting this to address separately.

stephenroller commented on parlai/tasks/cnn_dm/agents.py:
```diff
@@ -0,0 +1,91 @@
+        return os.path.join(self.datapath, dt + '.txt')
+
+    def setup_data(self, input_path):
+        def fix_missing_period(line):
```
Can we move this method outside? I don't think it has any closures.

stephenroller commented on parlai/tasks/cnn_dm/agents.py:
```diff
@@ -0,0 +1,91 @@
+                    missing_stories += 1
+                else:
+                    stories_added += 1
+                    with story_file:
```
what's this with do here?

stephenroller replied:
Can we refactor this whole method a bit?

stephenroller commented on parlai/tasks/cnn_dm/build.py:
```diff
@@ -0,0 +1,89 @@
+        build_data.make_dir(dpath)
+
+        # Download the data.
+        cnn_urls = {
```
let's make this a CNN_URLS constant for the whole file.

stephenroller commented on parlai/tasks/cnn_dm/build.py:
```diff
@@ -0,0 +1,89 @@
+        for dt in dm_urls:
+            (url, fname) = dm_urls[dt]
+            build_data.download(url, dpath, fname)
+            with open(os.path.join(dpath, fname)) as urls_file, open(
```
let's pull out the filename construction from this so the pair of opens fit on one line. (Whole construct will be 3 lines)

stephenroller commented on parlai/tasks/mwsc/agents.py:
```diff
@@ -0,0 +1,50 @@
+                answer = schema_line.get('answer')
+                question = schema_line.get('question')
+                context = schema_line.get('context')
+                yield (context + '\n' + question,
```
If this doesn't fit on one line (shouldn't it?e, I'd prefer
```
yield (
    context + '\n' + question,
    [answer],
    None,
    None
), new_episode
```

Although IIRC, the None's are redundant, and you can do:
```
yield (context + '\n' + question, [answer]), new_episode
```

stephenroller commented on parlai/tasks/mwsc/build.py:
```diff
@@ -0,0 +1,86 @@
+
+        # Download the data.
+        fname = 'schema.txt'
+        url = 'https://raw.githubusercontent.com/salesforce/decanlp/' \
```
Prefer
```
url = (
    'https:/...'
    'd5d'
)
```
but you can also do
```
url = 'https://.../' + fname  # noqa: E501
```

stephenroller commented on parlai/tasks/mwsc/build.py:
```diff
@@ -0,0 +1,86 @@
+        for split, examples in zip(splits, [train, dev, test]):
+            with open(os.path.join(dpath, f'{split}.json'), 'a') as split_file:
+                for ex in examples:
+                    split_file.write(json.dumps(ex)+'\n')
```
space around +

stephenroller commented on parlai/tasks/mwsc/build.py:
```diff
@@ -0,0 +1,86 @@
+
+        splits = ['train', 'validation', 'test']
+        for split, examples in zip(splits, [train, dev, test]):
+            with open(os.path.join(dpath, f'{split}.json'), 'a') as split_file:
```
I think we aimed to support Python 3.5 a bit more so we have to avoid f-strings 😢 (but I'm not certain we haven't already broken 3.5 support...)

stephenroller commented on parlai/tasks/qasrl/agents.py:
```diff
@@ -0,0 +1,92 @@
+            for qa in convert_to_qa(data):
+                question = qa[0]
+                answers = qa[1]
+                yield (question, answers, None, None), new_episode
```
Again I think we can drop the None's, but this is also good as is.

stephenroller commented on parlai/tasks/squad/agents.py:
```diff
@@ -196,10 +196,8 @@ def setup_data(self, path):
-                        '\n'.join([title, context, question]),
-                        answers
-                    ), True
+                    yield ('\n'.join([title, context, question]),
```
lol you don't like it?

hadasah replied:
lolll I don't think this was me. Not sure what the rationale was, so I'll change it back.

stephenroller commented on parlai/tasks/task_list.py:
```diff
@@ -718,6 +718,98 @@
             "random from the whole dataset."
         ),
     },
+    {
```
😍 soooo much addition.

Can we add "decanlp" as a tag to all of these please?

stephenroller commented on parlai/tasks/woz/build.py:
```diff
@@ -0,0 +1,33 @@
+# This source code is licensed under the MIT license found in the
+# LICENSE file in the root directory of this source tree.
+#
+# Download and build the data if it does not exist.
```
This comment seems a bit outdated or misplaced.

hadasah replied:
Do you mean "Download and build the data if it does not exist." ? That's at the top of all of the build.py files I think

stephenroller replied:
okay nvm i guess

jaseweston commented:
so if one wants to train on all the decanlp tasks at once, what is the
procedure to do that that:   -t "#decanlp" or something like that?    i
more imagined   "-t decanlp"  .. oh i guess we could do that by the decanlp
task just referencing all the other teachers... ?

On Tue, Mar 12, 2019 at 6:25 PM hadasah <notifications@github.com> wrote:

> @CharlieChenFB <https://github.com/CharlieChenFB> added the DecaNLP
> tasks. I've updated his PR so that the CNN_DM, MWSC, and SST datasets w/
> train/dev/test splits match up with the DecaNLP paper and repo.
> Two notable discrepancies with the DecaNLP tasks remain:
>
>    1. IWSLT in the DecaNLP task uses 2016 for train, 2013 and 2014 test
>    sets for dev and test, resp. I just kept IWSLT 2014 for now, we can add a
>    DecaNLP specific IWSLT task if we feel strongly.
>    2. WOZ # examples in the paper isn't what we get when running
>    display_data. Based on the repo, it seems they might use both English and
>    German data despite reporting only using English in the paper. I'm keeping
>    just English for and initial release, display_data output looks reasonable,
>    we can revisit if this is important.
>
> Only adding reviewers who responded to the original PR for now.
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/facebookresearch/ParlAI/pull/1546
> Commit Summary
>
>    - implemented basic functionality for decaNLP tasks
>    - implemented basic functionality for DecaNLP tasks
>    - pulled lastest origin master
>    - Merge branch 'master' of https://github.com/facebookresearch/ParlAI
>    - removed print from display_data
>    - fixed mismatched task name for cnn_dm
>    - fixed mismatched task name for cnn_dm
>    - added copyrights
>    - reformatted files
>    - more reformatting
>    - more reformatting to pass travis
>    - Changed the implementation of the DecaNLP Teacher to use the
>    MultiTaskTeacher
>    - Merge branch 'master' into master
>    - Recommitting the multi task teacher
>    - amending accidental change to iwslt14 build
>    - Merge branch 'master' into master
>    - implemented the datasplits correctly, added descriptions to tasks
>    - minor reformatting to subdue Travis
>    - minor reformatting to appease Travis
>    - minor reformatting to appease Travis
>    - removed extraneous code from sst teacher which is causing errors
>    - removed unnecessary deep copy opt in teachers
>    - remove space
>    - Merge https://github.com/facebookresearch/ParlAI
>    - Merge branch 'master' into master
>    - nit in file
>    - updating with master
>    - linter fix
>    - Amended valid/dev/validate confusion
>    - lint
>    - pulling from upstream
>    - reverting bad pull
>    - fixed recursion in multitaskteacher
>    - Merge branch 'master' into decanlp
>    - Added correct SST urls for DecaNLP task, updated setup_data
>    - Updated CNN_DM with correct splits and label parsing
>    - Add MWSC splits
>    - Update task list
>    - Appease linter
>    - Update copyrights, lint fix
>
> File Changes
>
>    - *M* parlai/core/agents.py
>    <https://github.com/facebookresearch/ParlAI/pull/1546/files#diff-0>
>    (7)
>    - *A* parlai/tasks/cnn_dm/__init__.py
>    <https://github.com/facebookresearch/ParlAI/pull/1546/files#diff-1>
>    (0)
>    - *A* parlai/tasks/cnn_dm/agents.py
>    <https://github.com/facebookresearch/ParlAI/pull/1546/files#diff-2>
>    (91)
>    - *A* parlai/tasks/cnn_dm/build.py
>    <https://github.com/facebookresearch/ParlAI/pull/1546/files#diff-3>
>    (89)
>    - *A* parlai/tasks/decanlp/__init__.py
>    <https://github.com/facebookresearch/ParlAI/pull/1546/files#diff-4>
>    (0)
>    - *A* parlai/tasks/decanlp/agents.py
>    <https://github.com/facebookresearch/ParlAI/pull/1546/files#diff-5>
>    (78)
>    - *M* parlai/tasks/iwslt14/agents.py
>    <https://github.com/facebookresearch/ParlAI/pull/1546/files#diff-6>
>    (1)
>    - *M* parlai/tasks/multinli/agents.py
>    <https://github.com/facebookresearch/ParlAI/pull/1546/files#diff-7>
>    (3)
>    - *A* parlai/tasks/mwsc/__init__.py
>    <https://github.com/facebookresearch/ParlAI/pull/1546/files#diff-8>
>    (0)
>    - *A* parlai/tasks/mwsc/agents.py
>    <https://github.com/facebookresearch/ParlAI/pull/1546/files#diff-9>
>    (50)
>    - *A* parlai/tasks/mwsc/build.py
>    <https://github.com/facebookresearch/ParlAI/pull/1546/files#diff-10>
>    (86)
>    - *A* parlai/tasks/qasrl/__init__.py
>    <https://github.com/facebookresearch/ParlAI/pull/1546/files#diff-11>
>    (0)
>    - *A* parlai/tasks/qasrl/agents.py
>    <https://github.com/facebookresearch/ParlAI/pull/1546/files#diff-12>
>    (92)
>    - *A* parlai/tasks/qasrl/build.py
>    <https://github.com/facebookresearch/ParlAI/pull/1546/files#diff-13>
>    (31)
>    - *A* parlai/tasks/qazre/__init__.py
>    <https://github.com/facebookresearch/ParlAI/pull/1546/files#diff-14>
>    (0)
>    - *A* parlai/tasks/qazre/agents.py
>    <https://github.com/facebookresearch/ParlAI/pull/1546/files#diff-15>
>    (57)
>    - *A* parlai/tasks/qazre/build.py
>    <https://github.com/facebookresearch/ParlAI/pull/1546/files#diff-16>
>    (35)
>    - *M* parlai/tasks/squad/agents.py
>    <https://github.com/facebookresearch/ParlAI/pull/1546/files#diff-17>
>    (12)
>    - *A* parlai/tasks/sst/__init__.py
>    <https://github.com/facebookresearch/ParlAI/pull/1546/files#diff-18>
>    (0)
>    - *A* parlai/tasks/sst/agents.py
>    <https://github.com/facebookresearch/ParlAI/pull/1546/files#diff-19>
>    (69)
>    - *A* parlai/tasks/sst/build.py
>    <https://github.com/facebookresearch/ParlAI/pull/1546/files#diff-20>
>    (38)
>    - *M* parlai/tasks/task_list.py
>    <https://github.com/facebookresearch/ParlAI/pull/1546/files#diff-21>
>    (92)
>    - *A* parlai/tasks/wikisql/__init__.py
>    <https://github.com/facebookresearch/ParlAI/pull/1546/files#diff-22>
>    (0)
>    - *A* parlai/tasks/wikisql/agents.py
>    <https://github.com/facebookresearch/ParlAI/pull/1546/files#diff-23>
>    (104)
>    - *A* parlai/tasks/wikisql/build.py
>    <https://github.com/facebookresearch/ParlAI/pull/1546/files#diff-24>
>    (37)
>    - *A* parlai/tasks/woz/__init__.py
>    <https://github.com/facebookresearch/ParlAI/pull/1546/files#diff-25>
>    (0)
>    - *A* parlai/tasks/woz/agents.py
>    <https://github.com/facebookresearch/ParlAI/pull/1546/files#diff-26>
>    (75)
>    - *A* parlai/tasks/woz/build.py
>    <https://github.com/facebookresearch/ParlAI/pull/1546/files#diff-27>
>    (33)
>
> Patch Links:
>
>    - https://github.com/facebookresearch/ParlAI/pull/1546.patch
>    - https://github.com/facebookresearch/ParlAI/pull/1546.diff
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/facebookresearch/ParlAI/pull/1546>, or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AKjk-AXfPQKHVgPeZidvOOxrcLXBjQasks5vWCligaJpZM4bsAYm>
> .
>

hadasah commented:
@jaseweston  either "-t #decanlp" or "-t decanlp" should work since the default DecaNLPTeacher is a multitaskteacher over all of the tasks, and I tagged all 10 subtasks with the 'decanlp' tag Stephen suggested. 

@stephenroller Left the extraneous Nones in place for now - I like them as a personal preference, but happy to get rid of them if you feel it's cleaner.

stephenroller commented on parlai/tasks/decanlp/agents.py:
```diff
@@ -64,13 +64,12 @@ class MwscTeacher(mwsc.DefaultTeacher):
                          'qasrl', 'qazre', 'woz', 'wikisql', 'mwsc']
         opt = deepcopy(opt)
-        opt['task'] = ', '.join(f'{task}' for task in decanlp_tasks)
+        opt['task'] = ', '.join(decanlp_tasks)
```
I think mayyyyyybe this can't have the space?

stephenroller commented on parlai/tasks/mwsc/build.py:
```diff
@@ -25,11 +25,13 @@ def build(opt):
         build_data.download(url, dpath, fname)
 
-        pattern = '\[.*\]'
+        pattern = '\\[.*\\]'
```
you can also do `pattern = r'\[.*\]'` I believe.

stephenroller commented on parlai/tasks/task_list.py:
```diff
@@ -768,7 +768,7 @@
         "display name": "WOZ restuarant reservation (Goal-Oriented Dialogue)",
         "task": "woz",
-        "tags": ["All"],
+        "tags": ["All", "decanlp", "Goal"],
```
nice addition

stephenroller commented:
(regarding the test, the error message is wrong, jack fixed it in #1553)

hadasah commented:
Flagging to @alexholdenmiller and @stephenroller: I introduced a newline to iwslt14 and accidentally discovered that iwslt14 fails the dataset test:

======================================================================
ERROR: test_verify_data (__main__.TestNewTasks)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/margaretli/ParlAI/parlai/core/agents.py", line 731, in _create_task_agents
    create_agent = getattr(my_module, 'create_agents')
AttributeError: module 'parlai.tasks.iwslt14.agents' has no attribute 'create_agents'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "tests/datatests/test_new_tasks.py", line 63, in test_verify_data
    text, log = verify(opt, print_parser=False)
  File "/Users/margaretli/ParlAI/parlai/scripts/verify_data.py", line 61, in verify
    world = create_task(opt, agent)
  File "/Users/margaretli/ParlAI/parlai/core/worlds.py", line 1029, in create_task
    world = create_task_world(opt, user_agents, default_world=default_world)
  File "/Users/margaretli/ParlAI/parlai/core/worlds.py", line 988, in create_task_world
    opt, user_agents, default_world=default_world
  File "/Users/margaretli/ParlAI/parlai/core/worlds.py", line 953, in _get_task_world
    task_agents = _create_task_agents(opt)
  File "/Users/margaretli/ParlAI/parlai/core/agents.py", line 735, in _create_task_agents
    return create_task_agent_from_taskname(opt)
  File "/Users/margaretli/ParlAI/parlai/core/agents.py", line 689, in create_task_agent_from_taskname
    task_agents = teacher_class(opt)
  File "/Users/margaretli/ParlAI/parlai/tasks/iwslt14/agents.py", line 34, in __init__
    source, target = task_name.split('_')
ValueError: not enough values to unpack (expected 2, got 1)

----------------------------------------------------------------------
Ran 1 test in 57.006s

FAILED (errors=1)

I got rid of the newline and am rerunning test_new_tasks.py

hadasah commented:
Output from test_new_tasks.py below:
(After removing the trivial changes made to squad and iwslt)
----------------------------------------------------------------------
Ran 1 test in 32180.001s

OK

The newline/tabbing changes to squad forced test_new_tasks to run on squad and led to the error below. It looks like this is also just an issue of squad being older than the test:

ERROR: test_verify_data (__main__.TestNewTasks)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/private/home/margaretli/ParlAI/parlai/core/agents.py", line 731, in _create_task_agents
    create_agent = getattr(my_module, 'create_agents')
AttributeError: module 'parlai.tasks.squad.agents' has no attribute 'create_agents'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "tests/datatests/test_new_tasks.py", line 63, in test_verify_data
    text, log = verify(opt, print_parser=False)
  File "/private/home/margaretli/ParlAI/parlai/scripts/verify_data.py", line 61, in verify
    world = create_task(opt, agent)
  File "/private/home/margaretli/ParlAI/parlai/core/worlds.py", line 1029, in create_task
    world = create_task_world(opt, user_agents, default_world=default_world)
  File "/private/home/margaretli/ParlAI/parlai/core/worlds.py", line 988, in create_task_world
    opt, user_agents, default_world=default_world
  File "/private/home/margaretli/ParlAI/parlai/core/worlds.py", line 953, in _get_task_world
    task_agents = _create_task_agents(opt)
  File "/private/home/margaretli/ParlAI/parlai/core/agents.py", line 735, in _create_task_agents
    return create_task_agent_from_taskname(opt)
  File "/private/home/margaretli/ParlAI/parlai/core/agents.py", line 689, in create_task_agent_from_taskname
    task_agents = teacher_class(opt)
  File "/private/home/margaretli/ParlAI/parlai/tasks/squad/agents.py", line 217, in __init__
    super().__init__(opt, shared)
  File "/private/home/margaretli/ParlAI/parlai/core/teachers.py", line 1144, in __init__
    self._setup_data(opt.get('parlaidialogteacher_datafile'))
  File "/private/home/margaretli/ParlAI/parlai/core/teachers.py", line 1170, in _setup_data
    with open(path) as read:
FileNotFoundError: [Errno 2] No such file or directory: '/private/home/margaretli/ParlAI/data/SQuAD-fulldoc/squad_fulldocs.train:ordered'
